ProjectSource/app-library/src/test/java/com/kayako/sdk/android/k5/kre/base/SocketConnectionTest.java

```diff
@@ -0,0 +1,230 @@
+import static org.mockito.ArgumentMatchers.contains;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.powermock.api.mockito.PowerMockito.whenNew;
+
+@PrepareForTest({
+        NetworkUtils.class,
+        Kayako.class,
+        TextUtils.class,
+        Uri.class,
+        Socket.class,
+        SocketConnection.class,
+        KayakoLogHelper.class,
+        KreConnection.class
+})
+public abstract class SocketConnectionTest {
+    private static final String INSTANCE_URL = "https://support.kayako.com";
+    private static final String SESSION_ID = "session_id";
+    private static final String REAL_TIME_URL = "realTimeUrl";
+    private static final String USER_AGENT = "userAgent";
+    private static final String FINGERPRINT_ID = "fingerprintId";
+    private static final String HOST = "host";
+    private static final String URL_STRING = "url_string";
+    private static final Throwable throwable = new Throwable();
+    private static final String NO_NETWORK_CONNECTION = "No Network Connection";
+    static final String ERROR_MESSAGE = "errorMessage";
+
+    @Mock
+    private SocketConnection.OnOpenSocketConnectionListener onOpenSocketConnectionListener;
+    @Mock
+    private SocketConnection.OnCloseConnectionListener onCloseConnectionListener;
+    @Mock
+    private Uri uri;
+    @Mock
+    private Uri.Builder builder;
+    @Mock
+    private Context context;
+    @Mock
+    Socket socket;
+    @Captor
+    ArgumentCaptor<ISocketOpenCallback> socketOpenCaptor;
+    @Captor
+    ArgumentCaptor<String> errorMessageCaptor;
+    @Captor
+    private ArgumentCaptor<ISocketCloseCallback> socketCloseCaptor;
+    @Captor
+    ArgumentCaptor<IErrorCallback> errorCaptor;
+    @Captor
+    private ArgumentCaptor<Socket.OnSocketThrowExceptionListener> socketThrowCaptor;
+    KreSessionCredentials kreCredentials;
+    private final SocketConnection socketConnection = new SocketConnection();
+
+    public void setUp() throws Exception {
+        mockStatic(NetworkUtils.class);
+        mockStatic(Kayako.class);
+        mockStatic(TextUtils.class);
+        mockStatic(Uri.class);
+        mockStatic(KayakoLogHelper.class);
+
+        whenNew(Socket.class).withArguments(URL_STRING).thenReturn(socket);
+        when(Kayako.getApplicationContext()).thenReturn(context);
+        when(NetworkUtils.isConnectedToNetwork(context)).thenReturn(true);
+        when(Uri.parse(or(eq(INSTANCE_URL), contains(REAL_TIME_URL)))).thenReturn(uri);
+        when(uri.getHost()).thenReturn(HOST);
+        when(uri.toString()).thenReturn(URL_STRING);
+        when(uri.buildUpon()).thenReturn(builder);
+        when(builder.build()).thenReturn(uri);
+        when(builder.appendQueryParameter(anyString(), anyString())).thenReturn(builder);
```

Comment:
I tried my best not to use `any()` or `anyXXX()` matchers but there are four lines that I had to resort to them
```
    when(builder.appendQueryParameter(anyString(), anyString())).thenReturn(builder);
    when(socket.onOpen(any(ISocketOpenCallback.class))).thenReturn(socket);
    when(socket.onClose(any(ISocketCloseCallback.class))).thenReturn(socket);
    when(socket.onError(any(IErrorCallback.class))).thenReturn(socket);
```
The first line is executed for five different arguments that is of no concern to this unit test and for the other three lines I had to use them because in the main code the arguments passed to these methods are anonymous inner classes (`ISocketOpenCallback`, `ISocketCloseCallback` and `IErrorCallback` are interfaces)